### ISSUE TITLE
Geometry_oM: adj Tolerance

### ISSUE THREAD

**Issue by michaldengusiak** [BHoM/BHoM#550]
#### Description:

I noticed that we have a defined tolerance (1e-6)  but we are missing a way to change it to different value for example from `Distance `to `MacroDistance `or adj `Distance `to a different value  for the user

```
namespace BH.oM.Geometry
{
    public static class Tolerance
    {
        /***************************************************/
        /**** Constants                                 ****/
        /***************************************************/

        public const double MicroDistance = 1e-9;
        public const double Distance = 1e-6;
        public const double MacroDistance = 1e-3;
        public const double Angle = 1e-6;

        /***************************************************/
    }
}
```

**Comment by FraserGreenroyd** [comment]
These are fixed constant values (keyword `const`) so should not be changed. All places where they are used in the code and in components allow users to change the input values because these form the basis of defaults. I.E. `public static double SomeMethod(this Polyline curve, double tolerance = Tolerance.Distance)` means users can use `SomeMethod` immediately without needing to worry about tolerance if they don't want to, but if they need to specify a different tolerance, the joy of the reflection into the UIs allows users to plug in an overriding tolerance specific for their needs.

Long story short, there is no need to change this file, nor bring in capabilities to allow users to change these values. Closing as non-issue.

**Comment by pawelbaran** [comment]
@FraserGreenroyd you just explained how it works now. And what @michaldengusiak means is how he would like it to work - I believe it would be worth considering to add an extra layer that would allow for overwriting the constants in certain domains (say, scripts/documents).

I can imagine this being very practical: having a single component that changes the tolerance globally could 1. save time needed for clicking through every component 2. enforce consistency across the whole workflow.

@al-fisher @rwemay @IsakNaslundBh what do you think?

**Comment by FraserGreenroyd** [comment]
I would argue that any change to tolerances should be done with care, especially given a change to a global tolerance could cause untold problems in other toolkits downstream. For example, a toolkit which relies on `BH.oM.Geometry.Tolerance.Distance` to be `1e-6`, to suddenly have that changed to `0.01` by a user could cause models to fall out of alignment by accident.

I would say that either it is better to have a series of datasets with reasonable, reviewed, tolerances that can serve as defaults, or, we keep the system as it currently is, whereby users can change tolerances when tolerances are optional inputs on components/methods, as this would be safer than changing a `const` oM property. Otherwise, this would need to add protections to a lot of toolkits to implement if we could not guarantee that `Tolerance.Distance` will remain the same.

That's my thoughts and I would reclose but am happy for a discussion to continue if a discussion is now needed, though if the discussion goes stale, I'll then reclose as a non-issue :wink:

**Comment by michaldengusiak** [comment]
I agree with Pawel...we need freedom. 

Currently, I run some test on my workflow and by globally changing tolerance all my script works perfectly. Otherwise, I need to debug and change tolerance in dozen of places. Now we need to find a neat way to implement or give a new option to users. 

I wish we have 3 options... 
1. standard default as we have now
2. global change for standard Tolerance value
3. all tolerances exposed so we can change in each step from GH, Dynamo level...tolerances

**Comment by FraserGreenroyd** [comment]
In that case the discussion should be broadened to include people like @IsakNaslundBh @alelom @epignatelli @ZiolkowskiJakub and @adecler and other people who use the tolerances. One workflow is not enough evidence to change something which had a decision to make `const` made against it. That said, I am not saying we should stick by whatever we have previously decided and we should always be free to discuss new options, but there should be evidence to show why the change is necessary, and the benefits it might bring. Equally, the downsides to it need to be evaluated.

Option 3 is the one we should be going for, and it is an unfortunate oversight that not all tolerances were exposed at the time the methods in the engines were added, but that is the far more acceptable (to me) fix than changing the oM itself, given the reasons I have already stated above.

As I say, happy to have the discussion, but I would want to see input from a broader range of people with a broader range of workflows discussed than just one before necessitating a change to this particular part of the oM.

**Comment by rwemay** [comment]
I think all tolerances should be exposed and not hardcoded in the central engine for complete transparency. The software/discipline/workflow specific defaults then should be set in the user interfaces (this is where templates/samples will be really useful), or toolkits (but again being careful not to hardcode too much/make it opaque - consider units/region specificity etc). 

It could be useful to have a system of global tolerance control, much like many of the UI's we're loading into. I'm not sure how the methods for individual flow-based workflows will call/refer to a global setting though.

**Comment by FraserGreenroyd** [comment]
Closing issue for tidiness due to the discussion going stale. From the discussion that occurred, the consensus is that tolerances should always be exposed where possible, but those exposures are changes for BHoM Engine and suitable issues should be raised to achieve this on a case-by-case basis (i.e. when we find a tolerance needing exposure, we expose it) rather than a large, catch-all issue which risks something being missed. Hope that makes sense :smile:
